MQTT: Support wildcards in topic filters matching retained messages

getlarge opened the pull request:
See description and conversation in https://github.com/rabbitmq/rabbitmq-server/pull/13048

getlarge committed:
test(rabbitmq_mqtt): add wildcard tests for retained messages in ETS store

getlarge committed:
feat(rabbitmq_mqtt): add has_wildcards function to check for wildcard topics

getlarge committed:
feat(rabbitmq_mqtt): add rabbit_globber module for wildcard topic matching

getlarge committed:
test(rabbitmq_mqtt): add unit tests for rabbit_globber module functionality

getlarge committed:
refactor(rabbitmq_mqtt): rename test functions for consistency and clarity

getlarge committed:
chore: rename rabbit_globber to rabbit_mqtt_topic_matcher

getlarge committed:
refactor(rabbitmq_mqtt): add retained message ets store structure and functionality

getlarge committed:
refactor(rabbitmq_mqtt): support sending retained messages list

getlarge committed:
test(rabbitmq_mqtt): update publish expectations

getlarge committed:
refactor(rabbitmq_mqtt): remove rabbit_mqtt_topic_matcher and associated files

getlarge committed:
refactor(rabbitmq_mqtt): support retained message list handling

getlarge committed:
fix(rabbitmq_mqtt): enhance ETS table management and recovery process

getlarge committed:
fix(rabbitmq_mqtt): replace ets:delete_object with ets:match_delete

getlarge committed:
test(rabbitmq_mqtt): update retained message tests and add recovery scenario

getlarge committed:
refactor(rabbitmq_mqtt): add support for wildcard filtering for DETS store

getlarge committed:
test(rabbitmq_mqtt): add wildcard filtering tests for DETS store

getlarge committed:
fix(rabbitmq_mqtt): improve topics deletion efficiency

getlarge committed:
refactor(rabbitmq_mqtt): improve root node handling and add custom file path in DETS store

getlarge committed:
test(rabbitmq_mqtt): ensure DETS tables are unique

getlarge committed:
feat(rabbitmq_mqtt): add max retained messages count configuration for MQTT plugin

getlarge committed:
ci: trigger

getlarge committed:
fix(mqtt): update topic handling in retained message processing and tests

getlarge committed:
test(mqtt): add should_remove_from_retained test for message retention behavior

robertsLando commented:
when I delete a retained message I see a publish of {} instead. After this if I disconnect and re-connect I don't see it anymore so it's partially working.

[Screencast da 2025-02-12 10-50-09.webm](https://github.com/user-attachments/assets/2d99b29d-54e5-487e-91e4-e4c9a73ea0fd)

mortdiggiddy commented:
Hello there Mr. getLarge.

I have followed your work regarding this ticket and all of the progress you made, right up until the point where it would not be accepted by the RabbitMQ team :)

We are very interested in a feature like this, I work in a Robotics Lab in Dubai. I ran into this issue the other day trying to understand why retained topics with a '+' for word replacement was not working when the retain flag is set by the publisher.

I have thoroughly reviewed the entire history of this request, all the way back to July 2018 where it seems to have started.

How can my team use your work and adopt this (Earlang plugin?) into our running RabbitMQ single instance as well as our 3.13 instances running inside the Rabbit MQ cluster operator on Kubernetes.

I am willing to help write any code to make this happen.

@getlarge @robertsLando

ansd committed:
Fix shutdown and type of child supervisor

Previously, the following three supervisors used the wrong `shutdown`
and wrong `type`:
* rabbit_exchange_federation_sup
* rabbit_federation_sup
* rabbit_queue_federation_sup

For `shutdown` Erlang/OTP recommends:
"If the child process is another supervisor, the shutdown time must be
set to infinity to give the subtree ample time to shut down. Setting the
shutdown time to anything other than infinity for a child of type supervisor
can cause a race condition where the child in question unlinks its own children,
but fails to terminate them before it is killed."

For `type` Erlang/OTP recommends:
"type specifies if the child process is a supervisor or a worker.
The type key is optional. If it is not specified, it defaults to
worker."

This commit fixes the wrong child spec by using a timeout of `infinity`
and type `supervisor`.

(cherry picked from commit cfcf6cffa2cc313c568e925d091c953b6f72618f)

mkuratczyk committed:
Check exclusive queue owner before deletion

[Why]
For a long time, there has been race condition when deleting
exclusive queues - if a connection was re-established and a queue
with the same name was declared, we could delete the new queue.

For example, with many MQTT consumers, if we performed a rolling restart
of the cluster and the clients reconnected without any delay, after the
restart, we sometimes had the expected number of connections but a lower
number of queues, even though there should be a queue for each consumer.

[How]
Check that the exclusive_owner has the value we expect when requesting
deletion. If the value is different, this means this is effectively a
different queue (same name, but a different connection), so we should
not delete it.

ansd committed:
Stop federation supervision hierachies before stopping `rabbit`

 ## What?

Federation links started in the federation plugins are put
under the `rabbit` app supervision tree (unfortunately).

This commit ensures that the entire federation supervision hierarchies
(including all federation links) are stopped **before** stopping app
`rabbit` when stopping RabbittMQ.

 ## Why?

Previously, we've seen cases where hundreds of federation links are
stopped during the shutdown procedure in app `rabbit` leading to
federation link restarts happening in parallel to vhosts being stopped.
In one case, the shutdown of app `rabbit` even got stuck (although there
is no evidence that federation was the problem).

Either way, the cleaner appraoch is to gracefully stop all federation
links, i.e. the entire supervision hierarchy under
`rabbit_exchange_federation_sup` and `rabbit_queue_federation_sup`
when stopping the federation apps, i.e. **before** proceeding to stop
app `rabbit`.

 ## How?

The boot step cleanup steps for the federation plugins are skipped when
stopping RabbitMQ.

Hence, this commit ensures that the supervisors are stopped in the
stop/1 application callback.

This commit does something similar to https://github.com/rabbitmq/rabbitmq-server/pull/14054
but uses a simpler approach.

(cherry picked from commit 8bffa5860d0d6b6275d3badf74352cac0f27e653)

michaelklishin committed:
Federation: disconnect links before stopping, in parallel

This yields very significant efficiency gains
with hundreds or thousands of links.

To avoid overwhelming the upstream schema data store
(which could be a 7-9 node cluster on 3.x with Mnesia),
we limit the degree of parallelism and add configurable
throttling delays into the process.

Technical design pair: @ansd.

michaelklishin committed:
Federation: use cuttlefish:unset/0 for new schema keys

without it, the new keys (or rather, their defaults) will spill into the `config_schema_SUITE`s of other plugins.

michaelklishin committed:
More rabbitmq.conf schema tests #15271

michaelklishin committed:
Integrate with maintenance mode #15271

We implement the `revive/0` part for symmetry. As with the revive command in general, it serves as a last resort available for rollback.

Usually nodes put into maintenance mode are shortly stopped for upgrading or reconfiguration.

michaelklishin committed:
Exclude these two from Dialyzer #15271

michaelklishin committed:
Federation: be more defensive during shutdown

when the core now interacts with a part of the
supervision tree owned by this plugin for
more efficient shutdown.

lhoguin committed:
CI: Remove unused shard5 node

lhoguin committed:
Merge pull request #15277 from rabbitmq/loic-rm-ct-shard5

CI: Remove unused shard5 node

mkuratczyk committed:
shovel: hopefully deflake amqp10_static suite (#15280)

reusing an auto-delete exchange is asking for trouble

michaelklishin committed:
Refactoring, logging tweaks #15271

michaelklishin committed:
Tweak link states further #15271

michaelklishin committed:
Extend #15276

michaelklishin committed:
Merge pull request #15271 from rabbitmq/mk-federation-parallel-disconnect

Federation: disconnect links before stopping, in parallel

michaelklishin committed:
Naming #15276

michaelklishin committed:
Federation: rework supervisor_shutdown_concurrency_safety to use await_condition/2

(cherry picked from commit 7c3f0392dd10af1dfab0fd58742bba87cc7edbdd)

michaelklishin committed:
Merge pull request #15276 from rabbitmq/check-owner-when-deleting-exclusive-queue

Check exclusive queue owner before deleting a queue

michaelklishin committed:
Drive-by change: remove an internals doc file from 2012

michaelklishin committed:
Federation: rework supervisor_shutdown_concurrency_safety further

Wait for both federation links to be established before proceeding
with expect_federation calls. The test was flaky because it declared
fed1.downstream2 inside the test function and immediately tried to
verify federation without waiting for the asynchronous link setup.

michaelklishin committed:
Federation: significant test suite refactoring

to reduce flakiness and speed test execution up.

michaelklishin committed:
Make sure that "Other ETS" and "Other system" memory metrics are never negative

They can become negative due to ETS tables being concurrently
added, removed and updated.

So simply report 0 in such cases.

michaelklishin committed:
Add unit_rabbit_vm to the CI group 3D

michaelklishin committed:
Refactor shovel tests the same way we did with federation in bfb3a722

michaelklishin committed:
Merge pull request #15289 from rabbitmq/mk-negative-ets-table-and-system-memory-metrics

Make sure that "Other ETS" and "Other system" memory metrics are never reported as negative

michaelklishin committed:
Further shovel test suites refactoring

ansd committed:
Speed up tests

```
rabbit_ct_broker_helpers:stop_broker/2
rabbit_ct_broker_helpers:start_broker/2
```

is about 6 seconds faster compared to

```
rabbit_ct_broker_helpers:stop_node/2
rabbit_ct_broker_helpers:start_node/2
```

This commit uses the faster versions in many tests where for example we
only want to trigger a leader change or a queue becoming unavailable.

In some test cases, we actually want the entire Erlang node to be down.
Those tests continue to use `rabbit_ct_broker_helpers:stop_node/2`.

michaelklishin committed:
Merge pull request #15292 from rabbitmq/da-stop-start-broker

Speed up tests

michaelklishin committed:
4.2.3 release notes updates

michaelklishin committed:
4.1.8 release notes updates

ansd committed:
Add test enabling all feature flags with exchange logger

Add a test enabling all stable feature flags with exchange logging enabled.

This is a more general test case for https://github.com/rabbitmq/rabbitmq-server/discussions/11652
that should catch all future feature flag issues when `rabbit_ff_controller` publishes a message.

Relates
https://github.com/rabbitmq/rabbitmq-server/discussions/11652
https://github.com/rabbitmq/rabbitmq-server/discussions/14069
https://github.com/rabbitmq/rabbitmq-server/pull/14796

michaelklishin committed:
Merge pull request #15294 from rabbitmq/ff-exchange-logger

Add test enabling all feature flags with exchange logger

michaelklishin committed:
xref: ignore a few federation plugin functions conditionally used in the core

mortdiggiddy commented:
Hey there friend. Got a notification about this PR in email automation.

Is this work revitalized?

Hallelujah if so.

getlarge commented:
@mortdiggiddy I will soon get in touch with the RabbitMQ maintainers and see if they have plan to improve the retained messages store for MQTT plugin.
Until then, this fork and branch are still alive, we are using them in our custom Docker image for RabbitMQ.
We also use this [new plugin](https://github.com/getlarge/rabbitmq-mqtt-annotation-interceptor) to add extra metadata about the MQTT clients.
